Hashing logic for c10::complex

anjali411 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #51472 Add support for complex valued keys for dict in TS
* **#51441 Hashing logic for c10::complex**

Differential Revision: [D26170195](https://our.internmc.facebook.com/intern/diff/D26170195)

anjali411 committed:
Hashing logic for c10::complex

[ghstack-poisoned]

anjali411 committed:
Update on "Hashing logic for c10::complex"




[ghstack-poisoned]

ezyang commented on c10/test/util/complex_test_common.h:
```diff
@@ -167,6 +169,11 @@ TEST(TestConstructors, FromThrust) {
+TEST(TestConstructors, UnorderedMap) {
+  std::unordered_map<c10::complex<double>, c10::complex<double>> m;
+  m[c10::complex<double>(2.5, 3)] = c10::complex<double>(2, -3.2);
+  m[c10::complex<double>(2, 0)] = c10::complex<double>(0, -3);
```
This isn't a particularly robust test haha

anjali411 replied:
lol yeah I updated the test to access values and check equality

anjali411 committed:
Update on "Hashing logic for c10::complex"


Differential Revision: [D26170195](https://our.internmc.facebook.com/intern/diff/D26170195)

[ghstack-poisoned]

zasdfgbnm commented on c10/test/util/complex_test_common.h:
```diff
@@ -167,6 +169,17 @@ TEST(TestConstructors, FromThrust) {
 }
 #endif
 
+TEST(TestConstructors, UnorderedMap) {
```
Why this test is part of `TestConstructors`? Hashing sounds unrelated to "constructor".

anjali411 replied:
yeah I guess the context in which these tests were named were for testing the construction of `c10::complex`, however it seemed to me that testing construction of containers containing `c10::complex` could fall under this category as well.

anjali411 committed:
Update on "Hashing logic for c10::complex"


Differential Revision: [D26170195](https://our.internmc.facebook.com/intern/diff/D26170195)

[ghstack-poisoned]
